src/mlpack/tests/callback_test.cpp

```diff
@@ -0,0 +1,118 @@
+
+#include <ensmallen.hpp>
+#include <ensmallen_bits/callbacks/callbacks.hpp>
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/rnn.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace mlpack;
+using namespace mlpack::ann;
+
+BOOST_AUTO_TEST_SUITE(CallbackTest);
+
+/**
+ * Test a FFN model with PrintLoss callback
+ */
+BOOST_AUTO_TEST_CASE(FFNCallbackTest)
+{
+  arma::mat data;
+  arma::mat labels;
+
+  data::Load("lab1.csv", data, true);
+  data::Load("lab3.csv", labels, true);
+
+  FFN<MeanSquaredError<>, RandomInitialization> model;
+
+  model.Add<Linear<>>(1, 2);
+  model.Add<SigmoidLayer<>>();
+  model.Add<Linear<>>(2, 1);
+  model.Add<SigmoidLayer<>>();
+
+  std::stringstream stream;
+  model.Train(data, labels, ens::PrintLoss(stream));
+
+  BOOST_REQUIRE_GT(stream.str().length(), 0);
+}
+
+/**
+ * Test a FFN model with PrintLoss callback and optimizer parameter
+ */
+BOOST_AUTO_TEST_CASE(FFNWithOptimizerCallbackTest)
+{
+  arma::mat data;
+  arma::mat labels;
+
+  data::Load("lab1.csv", data, true);
+  data::Load("lab3.csv", labels, true);
+
+  FFN<MeanSquaredError<>, RandomInitialization> model;
+
+  model.Add<Linear<>>(1, 2);
+  model.Add<SigmoidLayer<>>();
+  model.Add<Linear<>>(2, 1);
+  model.Add<SigmoidLayer<>>();
+
+  std::stringstream stream;
+  ens::StandardSGD opt(0.1, 1, 5);
+  model.Train(data, labels, ens::PrintLoss(stream));
```

Comment:
Should we pass opt here ? or did i miss something

Reply:
There are 2 Train methods defined - one with an explicit optimizer passed and one without.

Reply:
> There are 2 Train methods defined - one with an explicit optimizer passed and one without.

Then line no 59 is not needed, since you didn't use opt anywhere 

And also 35 and 60, both lines don't have an optimizer passed, line 86 and 113 is different, since we pass opt in Line 113 explicitly.

Reply:
Sorry my bad, I should have passed opt to the Train method. What a stupid mistake